rest/endpoint/v1/api.go

```diff
@@ -0,0 +1,874 @@
+		}
+	}
+
+	if !found {
+		RespondWithError(w, fmt.Errorf("column '%s' not found in table", columnName), http.StatusNotFound)
+		return
+	}
+
+	RespondJSONObjectWithCode(w, http.StatusOK, column)
+}
+
+func (s *routeList) AddColumn(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	keyspaceName := s.params(r, keyspaceParam)
+	tableName := s.params(r, tableParam)
+	user := auth.ContextUserOrRole(r.Context())
+
+	var columnDefinition m.ColumnDefinition
+	if err := parseAndValidatePayload(&columnDefinition, r); err != nil {
+		msg := "unable to parse payload"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "table", tableName)
+		RespondWithError(w, errors.New(msg), http.StatusBadRequest)
+		return
+	}
+
+	column, err := m.ToDbColumn(columnDefinition)
+
+	if err != nil {
+		RespondWithError(w, err, http.StatusBadRequest)
+	}
+
+	tableInfo := db.AlterTableAddInfo{
+		Keyspace: keyspaceName,
+		Table:    tableName,
+		ToAdd:    []*gocql.ColumnMetadata{column},
+	}
+
+	err = s.dbClient.AlterTableAdd(&tableInfo, newDbOptions(user))
+	if err != nil {
+		msg := "unable to execute alter table query"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "table", tableName, "error", err)
+
+		switch err.(type) {
+		case *e.ConflictError:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusConflict)
+			return
+		default:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	RespondJSONObjectWithCode(w, http.StatusCreated, m.TablesResponse{Success: true})
+}
+
+func (s *routeList) DeleteColumn(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	keyspaceName := s.params(r, keyspaceParam)
+	tableName := s.params(r, tableParam)
+	columnName := s.params(r, "columnName")
+	user := auth.ContextUserOrRole(r.Context())
+
+	err := s.dbClient.AlterTableDrop(&db.AlterTableDropInfo{
+		Keyspace: keyspaceName,
+		Table:    tableName,
+		ToDrop:   []string{columnName},
+	}, newDbOptions(user))
+
+	if err != nil {
+		msg := "unable to execute alter table query"
+		s.logger.Debug(msg, "keyspace", keyspaceName, "table", tableName, "column", columnName, "error", err)
+
+		switch err.(type) {
+		case *e.NotFoundError:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusNotFound)
+			return
+		default:
+			RespondWithError(w, fmt.Errorf(msg), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	RespondJSONObjectWithCode(w, http.StatusNoContent, nil)
+}
+
+func (s *routeList) GetRow(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	keyspaceName := s.params(r, keyspaceParam)
+	tableName := s.params(r, tableParam)
+	rowIdentifier := s.params(r, "rowIdentifier")
+	user := auth.ContextUserOrRole(r.Context())
+
+	// grab table and extract primary key col names
+	columns, values, err := primaryKeyValues(rowIdentifier, keyspaceName, tableName, s.dbClient)
```

Comment:
I'm concerned about `rowIdentifier`, specially regarding escaping.

I think we should pursue a single `<table>/query` route, using `GET` and url encoded query string, wdyt?

Should we create a separate ticket to do it?

Reply:
Doing so breaks CRUD and the current contract. As long as the identifier is URL encoded I don't see the problem as is.

Reply:
what contract? We just released a day ago. :)

Reply:
https://github.com/riptano/cloud-ondemand/blob/master/data-endpoint/api/swagger-v1.yaml#L851

Reply:
Its true that the semicolon is a valid char in the sub-delimiter category (similar to `&` and `?`), but having to send the parameters in that way is counterintuitive/error prone for users.

Http client apis usually have mechanisms to pass query string parameters as a dictionary and the library will encode them.

For example, using [jQuery ajax](https://api.jquery.com/jquery.ajax/#jQuery-ajax-settings):

```javascript
$.ajax({
  url: "v1/keyspaces/ks1/tables/tbl1/rows",
  type: "get",
  data: { 
    key1: "abc", 
    key2: "zz&1233",
  });
```

Will result in `v1/keyspaces/ks1/tables/tbl1/rows?key1=abc&key2=zz%261233`.

The same applies to other http client libraries in different lang/tech stacks. Having a custom delimiter format forces users to encode each parameter individually and join strings themselves.

Reply:
Ah, here's the docs I was using: https://astra.readme.io/reference#getrows-1. I'm unable to find a reference to the `;` deliminator there. Are those the external facing docs?

Reply:
Using query params is something we discussed during the initial design and implementation but decided against it because it's not ReSTful. This endpoint is identifying a specific resource whereas query params are used for search or filter operations. And I don't believe you have to encode then join, I believe you can join and then encode. Something like

```python
import urllib.parse

params = ["foo", "bar", "buzz"]

print(urllib.parse.quote_plus(";".join(params)))
```

@mpenick I didn't realize readme wasn't rendering the descriptions. I'll work on getting those to show up.

Reply:
>  I believe you can join and then encode

If you encode including the separators, there would be no difference between the character `;` in a value and the parameter delimiter.